File: trunk/hamcrest-python/hamcrest/library/number/ordering_comparison.py
from hamcrest.core.base_matcher import BaseMatcher


def _comparison(compare):
    if compare > 0:
        return 'less than'
    elif compare == 0:
        return 'equal to'
    else:
        return 'greater than'


class OrderingComparison(BaseMatcher):
    """Is the value a number equal to a value within some range of acceptable error?"""
    
    def __init__(self, value, min_compare, max_compare):
        self.value = value
        self.min_compare = min_compare
        self.max_compare = max_compare

    def matches(self, item):
        compare = cmp(self.value, item)
        return self.min_compare <= compare and compare <= self.max_compare

    def describe_to(self, description):
        description.append_text('a value ').append_text(_comparison(self.min_compare))
        if self.min_compare != self.max_compare:
            description.append_text(' or ').append_text(_comparison(self.max_compare))
        description.append_text(' ').append_value(self.value)



def greaterthan(value):
    return OrderingComparison(value, -1, -1)

def greaterthan_or_equalto(value):
    return OrderingComparison(value, -1, 0)

def lessthan(value):
    return OrderingComparison(value, 1, 1)

def lessthan_or_equalto(value):
    return OrderingComparison(value, 0, 1)

File: trunk/hamcrest-java/hamcrest-library/src/main/java/org/hamcrest/number/OrderingComparison.java
package org.hamcrest.number;

import org.hamcrest.Description;
import org.hamcrest.Factory;
import org.hamcrest.Matcher;
import org.hamcrest.TypeSafeMatcher;

public class OrderingComparison<T extends Comparable<T>> extends TypeSafeMatcher<T> {
    private final T value;
    private final int minCompare, maxCompare;

    private OrderingComparison(T value, int minCompare, int maxCompare) {
        this.value = value;
        this.minCompare = minCompare;
        this.maxCompare = maxCompare;
    }

    @Override
    public boolean matchesSafely(T other) {
        int compare = Integer.signum(value.compareTo(other));
        return minCompare <= compare && compare <= maxCompare;
    }

    public void describeTo(Description description) {
        description.appendText("a value ").appendText(comparison(minCompare));
        if (minCompare != maxCompare) {
            description.appendText(" or ").appendText(comparison(maxCompare));
        }
        description.appendText(" ").appendValue(value);
    }

    private String comparison(int compare) {
        if (compare > 0) {
            return "less than";
        }
        else if (compare == 0) {
            return "equal to ";
        }
        else {
            return "greater than";
        }
    }

    /**
     * Is value = expected?
     */
    @Factory
    public static <T extends Comparable<T>> Matcher<T> comparesEqualTo(T value) {
        return new OrderingComparison<T>(value, 0, 0);
    }

    /**
     * Is value > expected?
     */
    @Factory
    public static <T extends Comparable<T>> Matcher<T> greaterThan(T value) {
        return new OrderingComparison<T>(value, -1, -1);
    }

    /**
     * Is value >= expected?
     */
    @Factory
    public static <T extends Comparable<T>> Matcher<T> greaterThanOrEqualTo(T value) {
        return new OrderingComparison<T>(value, -1, 0);
    }

    /**
     * Is value < expected?
     */
    @Factory
    public static <T extends Comparable<T>> Matcher<T> lessThan(T value) {
        return new OrderingComparison<T>(value, 1, 1);
    }

    /**
     * Is value <= expected?
     */
    @Factory
    public static <T extends Comparable<T>> Matcher<T> lessThanOrEqualTo(T value) {
        return new OrderingComparison<T>(value, 0, 1);
    }
}